Go-to line, AI pathfinding modifications

stavrosfa opened the pull request:
1. Refactored the go-to layer, to show path lines between tiles.
2. The pathing algorith has been modified, now AI can see the whole map (as before) but the human player can't.
3. Account for narrow land strips in map for water units (see before below)

<img width="564" height="453" alt="c7_sea_travel" src="https://github.com/user-attachments/assets/42f6e4b3-23d2-4233-a69e-fadea6014a8e" />

and afer

<img width="548" height="410" alt="image" src="https://github.com/user-attachments/assets/47c46a08-c8ba-45a5-8fd7-494c063b7258" />


5. Grid layer is now computed by itself, before it was rendering only known tiles, resulting in missing lines etc. The tiles for the grid are already being computed for the other layers, so it's not like we add a huge overhead or something.

I wanted to include a "fixed" cursor animation, but I can't get it to render properly (civil disorder suffers too), and I 've tried for like hours and hours.
I don't know if it's the shadows, or the tint, or the overall way it's being extracted from the flc.
Someone smarter than me can understand it and fix it.

stavrosfa committed:
1. Refactored the go-to layer, to show path lines between tiles.
2. The pathing algorith has been modified, now AI can see the whole map (as before) but the human player can't.
3. Account for narrow land strips in map for water units
4. Grid layer is now computed by itself, before it was rendering only known tiles, resulting in missing lines etc. The tiles for the grid are already being computed for the other layers, so it's not like we add a huge overhead or something.

stavrosfa committed:
Whitespace formatting

stavrosfa committed:
1. Fixed unit tests
2. Added a new MapBase class to stash away functionlity we can use in multiple places
3. Added human factor in EdgeWalker
4. Added namespaces to the rest of the tests

TomWerner commented on C7Engine/AI/Pathing/AStarAlgorithm.cs:
```diff
@@ -76,6 +81,10 @@ public override TilePath PathFrom(Tile start, Tile destination) {
 
+					// TODO: the possibility here is to create a building/improvement/tech effect like "canal"
+					// (much like what the Panama canal is irl) that allows that kind of movement
+					if (unit.IsWaterUnit() && IsLandStrip(current, neighbor)) continue;
```
Does this break canal cities?

TomWerner replied:
Edit, I checked basic canals and they work, but would they work in the case from the land strip function?

stavrosfa replied:
If by canal cities you mean a city a water unit can go through, no it doesn't.

A simple one

<img width="665" height="276" alt="image" src="https://github.com/user-attachments/assets/cfae7fe1-729c-4586-8003-428dff024c20" />

And another, where the boat would never be able to get to because of the new restrictions.

<img width="665" height="346" alt="image" src="https://github.com/user-attachments/assets/fb85f1b9-a274-46d8-8ac6-40732d2d37cf" />


The LandStrip() method is only for sea-to-sea movement anyway

TomWerner commented on C7Engine/AI/Pathing/EdgeWalker.cs:
```diff
@@ -14,23 +14,24 @@ public UnitWalker(MapUnit unit) {
-				if (landUnit && !neighbor.IsLand()) {
-					continue;
+				if (unit.owner.isHuman && !unit.owner.HasExploredTile(neighbor)) {
+					isPassable = true;
```
This bit is very cool and faithful to the original game - could you call it out in the PR description?

stavrosfa replied:
Will do, thanks!

Actually, doesn't point no. 2 cover it?

TomWerner commented on C7/Map/GotoLayer.cs:
```diff
@@ -73,15 +75,49 @@ public override void drawObject(LooseView looseView, GameData gameData, Tile til
+					&& looseView.tileCenters.TryGetValue(nextTile, out Vector2 nextTileCenter)) {
+					staticCursorRect?.Hide();
+					looseView.DrawLine(currentTileCenter, nextTileCenter, Colors.Red, width: lineWidth);
+					// drawGotoCursor(looseView, nextTileCenter, gotoInfo.moveCost, gotoInfo.attackingMove);
```
Remove?

TomWerner reviewed (approved):
Thanks for the PR!

I did a couple of play tests and read most of the code, but I'll confess I didn't fully digest all of the map view/gotolayer changes.

On that note, it can be nice sometimes to have more focused PRs, for ease of reviewing/culprit finding. I think the pathfinding changes could probably have been their own PR (not a big deal at all - I'd rather have this great PR as one unit than no PR at all, but just some general feedback).

I found that it's significantly easier to do this with https://github.com/jj-vcs/jj, since `jj split` lets you very easily break up one commit into two stacked ones.

stavrosfa committed:
1. Freshwater continents (lakes) are now calculated taking into account the land strip restriction.
2. Added restriction to not draw the cursor when trying to pathfind with no unit selected
3. AI only will stop early trying to find a path from a continent to another
4. Don't try to draw anything in go-to layer when no unit is selected
5. Added unit tests for A* pathfinding and Continent calculations, updated EdgeWalker tests to use the new methods and variables in MapBase.cs
6. Moved the IsLandStrip() method in GameMap
